### PR TITLE
[torch-mlir][sparse] support e2e sparse kernels with COO inputs.

### PR DESCRIPTION
**Pull request by PeimingLiu**
**DO NOT MERGE!**: Still waiting llvm-project be synced with https://github.com/llvm/llvm-project/pull/82449

### CODE REVIEW

File: test/python/fx_importer/sparse_test.py [hunk 1]
```diff
@@ -158,7 +158,14 @@ def sparse_jit(f, *args, **kwargs):
     for a in args:
         if a.layout is torch.sparse_coo:
             xargs.append(a.values().numpy())
-            xargs.append(a.indices().numpy())
+            # Construct the additional position array required by MLIR with data
+            # array([0, nnz]).
+            nnz = a.indices().shape[1]
```
Comment (aartbik) [inline test/python/fx_importer/sparse_test.py:163]: I have been puzzled why torch.sparse does not just expose the nnz() as a method :-)

File: test/python/fx_importer/sparse_test.py [hunk 2]
```diff
@@ -158,7 +158,14 @@ def sparse_jit(f, *args, **kwargs):
...
+            xargs.append(torch.tensor([0, nnz], dtype=a.indices().dtype).numpy())
```
Comment (aartbik) [inline test/python/fx_importer/sparse_test.py:164]: why not do this directly in numpy?

File: test/python/fx_importer/sparse_test.py [hunk 3]
```diff
@@ -247,14 +258,17 @@ def forward(self, x, y):
     print(m)
 
     # Run it with PyTorch torch.sparse and with TORCH-MLIR sparse_jit.
-    # TODO: run with COO, right now only CSR works
     sparse_input = dense_input.to_sparse_csr()
```
Comment (aartbik) [inline test/python/fx_importer/sparse_test.py:261]: remove this csr() and res3

(that was the original intent of this test, i.e. run it only in the way we CHECK the IR above)

**Review by aartbik — COMMENTED** [review]
Make sure MLIR bumps to right hash first

File: test/python/fx_importer/sparse_test.py [hunk 4]
```diff
@@ -158,7 +158,14 @@ def sparse_jit(f, *args, **kwargs):
...
             xargs.append(a.values().numpy())
-            xargs.append(a.indices().numpy())
+            # Construct the additional position array required by MLIR with data
+            # array([0, nnz]).
+            nnz = a.indices().shape[1]
+            xargs.append(torch.tensor([0, nnz], dtype=a.indices().dtype).numpy())
+            # Transform a tensor<ndim x nnz> into [tensor<nnz> x ndim] to conform
+            # MLIR SoA COO representation.
+            for idx in a.indices():
+                xargs.append(idx.numpy())
```
Comment (aartbik) [inline test/python/fx_importer/sparse_test.py:168]: can this be done with a view change and then the append? also, does this generalize to n-dim COO?

Comment (PeimingLiu) [reply test/python/fx_importer/sparse_test.py:168]: I am hoping that it is already a view change (but I am not sure).

I think it generalize to n-dim COO. Per torch.sparse documentation " the indices of specified elements are collected in indices tensor of size (ndim, nse)"

File: test/python/fx_importer/sparse_test.py (hunk 1 above)
Comment (aartbik) [reply test/python/fx_importer/sparse_test.py:163]: actually _nnz() seems to work ;-)

File: test/python/fx_importer/sparse_test.py (hunk 2 above)
Comment (PeimingLiu) [reply test/python/fx_importer/sparse_test.py:164]: Per offline discussion, we will keep it this way to avoid clumsy conversion between `torch.dtype` to `numpy.dtype`

File: test/python/fx_importer/sparse_test.py (hunk 4 above)
Comment (PeimingLiu) [reply test/python/fx_importer/sparse_test.py:168]: I dumped the `data_ptr()` of `a.indices()` and `idx` used in the loop and confirm that they are alias.

File: test/python/fx_importer/sparse_test.py [hunk 5]
```diff
@@ -247,8 +253,7 @@ def forward(self, x, y):
     print(m)
 
     # Run it with PyTorch torch.sparse and with TORCH-MLIR sparse_jit.
-    # TODO: run with COO, right now only CSR works
-    sparse_input = dense_input.to_sparse_csr()
+    sparse_input = dense_input.to_sparse_coo()
```
Comment (aartbik) [inline test/python/fx_importer/sparse_test.py:256]: remove L256 completely, it is already in L251
(that way, it has the same structure as the first test above)

**Comment by aartbik** [comment]
bumped, please update your commit message before submitting

**Review by aartbik — APPROVED** [review]
